src/lib/oauth2-service.ts

```diff
@@ -206,59 +232,88 @@ class OAuth2Service extends EventEmitter {
-     * @param {object} response.body The body of the response.
-     * @param {Number} response.statusCode The HTTP status code of the response.
-     * @param {http.IncomingMessage} req The incoming HTTP request.
+     * @param {MutableResponse} response The response body and status code.
+     * @param {IncomingMessage} req The incoming HTTP request.
      */
     this.emit('beforeResponse', tokenEndpointResponse, req);
 
-    return res.status(tokenEndpointResponse.statusCode).json(tokenEndpointResponse.body);
-  }
+    return res
+      .status(tokenEndpointResponse.statusCode)
+      .json(tokenEndpointResponse.body);
+  };
 
-  [authorizeHandler](req, res) {
+  private authorizeHandler: RequestHandler = (req, res) => {
     const { scope, state } = req.query;
     const responseType = req.query.response_type;
     const redirectUri = req.query.redirect_uri;
     const code = uuidv4();
 
+    let queryNonce: string | undefined;
+
+    if ('nonce' in req.query) {
+      assertIsString(req.query.nonce, 'Invalid nonce type');
+      queryNonce = req.query.nonce;
+    }
+
+    assertIsString(redirectUri, 'Invalid redirectUri type');
+    assertIsString(scope, 'Invalid scope type');
+    assertIsString(state, 'Invalid state type');
+
     let targetRedirection;
     if (responseType === 'code') {
-      if (req.query.nonce) {
-        this[nonce][code] = req.query.nonce;
+      if (queryNonce !== undefined) {
+        this.#nonce[code] = queryNonce;
       }
-      targetRedirection = `${redirectUri}?code=${encodeURIComponent(code)}&scope=${encodeURIComponent(scope)}&state=${encodeURIComponent(state)}`;
+      targetRedirection = `${redirectUri}?code=${encodeURIComponent(
+        code
+      )}&scope=${encodeURIComponent(scope)}&state=${encodeURIComponent(state)}`;
     } else {
-      targetRedirection = `${redirectUri}?error=unsupported_response_type&error_description=The+authorization+server+does+not+support+obtaining+an+access+token+using+this+response_type.&state=${encodeURIComponent(state)}`;
+      targetRedirection = `${redirectUri}?error=unsupported_response_type&error_description=The+authorization+server+does+not+support+obtaining+an+access+token+using+this+response_type.&state=${encodeURIComponent(
+        state
+      )}`;
     }
 
-    res.redirect(targetRedirection);
-  }
-
-  [userInfoHandler](req, res) {
-    const userInfoResponse = {
+    // Note: This is a textbook definition of an "open redirect" vuln
+    // cf. https://cwe.mitre.org/data/definitions/601.html
+    //
+    // However, this whole library is expected to be used as a test helper,
+    // so there's no real point in making the exposed API more complex (by
+    // exposing an endpoint to preregister whitelisted urls, for instance)
+    // for the sake of security.
+    //
+    // This is *not* a real oAuth2 server. This is *not* to be run in production.
+    res.redirect(targetRedirection); // lgtm[js/server-side-unvalidated-url-redirection]
```

Comment:
@poveden This should silence 
![image](https://user-images.githubusercontent.com/92363/95622993-d91d6b00-0a74-11eb-8655-93e5357bbded.png)